### PR TITLE
Create a derive macro for HashStable and allow proc macros in rustc

### PR DESCRIPTION
**Pull request by Zoxc**
A combination of https://github.com/rust-lang/rust/pull/56864 and https://github.com/rust-lang/rust/pull/56795. There were complications with using `serde_derive` as rustc doesn't know which crate to use for the host when there is a serde_derive in the sysroot and cargo passes another on the command line built from crates.io.

r? @eddyb (for proc macro changes) @alexcrichton (for rustbuild changes) @michaelwoerister (for the macro itself)

### CODE REVIEW

**Comment by Zoxc** [comment]
Blocked on https://github.com/rust-lang/cargo/pull/6547

**Comment by michaelwoerister** [comment]
The macro looks awesome. This should let us get rid of so much boilerplate. Also, the `synstructure` crate looks pretty sweet `:)`

**Comment by alexcrichton** [comment]
I haven't been following this too too closely, but @Zoxc can you detail a bit more what some of the changes are to linking and such?

**Comment by Zoxc** [comment]
I'm just going to list the changes enabling proc macros.
- We pass `-Zdual-proc-macros` to both cargo and rustc when cross compiling. Cargo will then build proc macros both for the host and the target and pass both to rustc for crates which depend upon them. Rustc will link to the target proc macro crate in metadata, but load the host proc macro crate and use that to execute the macros.
- Bootstrap will now move proc macro crates compiled to the host into the sysroot at `lib\rustlib\<host-triple>\lib` in addition to the target crates which are moved to `lib\rustlib\<target-triple>\lib`
- We require the `Test` target for the compiler, since that compiles `libproc_macro` and we need that to build proc macros used by the stage1 compiler to build stage2 compiler crates.
- We make `-Z force-unstable-if-unmarked` affect proc macros too, to avoid an ICE in stability checking

File: src/librustc_macros/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "rustc_macros"
+version = "0.1.0"
+authors = ["The Rust Project Developers"]
```
Comment (mark-i-m) [inline src/librustc_macros/Cargo.toml]: Is it possible to make this a rust 2018 crate? Otherwise, somebody will come along and do that later for #58099 anyway, leading to more churn...

**Comment by alexcrichton** [comment]
FWIW I don't think I'll have much to add any more to this, so long as it works and bootstraps seems fine by me, I wouldn't block on me for this.

**Comment by Zoxc** [comment]
We'll need to get https://github.com/rust-lang/cargo/pull/6547 into beta before this can be merged.

**Comment by vext01** [comment]
This PR is now unblocked :)

**Comment by mati865** [comment]
@vext01 it's not that easy.
Updated Cargo has to make it's way to beta branch first, then bootstrap compiler has to updated.
After it's done this PR will be unblocked.

**Comment by Zoxc** [comment]
Ping @eddyb for review of the proc macro parts.

**Comment by Zoxc** [comment]
@rust-lang/compiler Do we have someone else who can review the proc macro changes in rustc here?

File: src/librustc_macros/src/hash_stable.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+use synstructure;
+use syn::{self, Meta, NestedMeta};
+use proc_macro2::{self, Ident, Span};
+
+struct Attributes {
+    ignore: bool,
+    project: Option<String>,
```
Comment (oli-obk) [inline src/librustc_macros/src/hash_stable.rs]: If you made this an `Ident`, would everything still work out? Then diagnostics would directly link to the right place.

File: src/librustc_macros/Cargo.toml (hunk 1 above)
Comment (oli-obk) [reply src/librustc_macros/Cargo.toml]: we could/should enforce this via tidy once we have all of them ported

File: src/librustc_macros/src/hash_stable.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    for attr in &field.attrs {
+        if let Ok(meta) = attr.parse_meta() {
+            if &meta.name().to_string() != "stable_hasher" {
+                continue;
+            }
+            let mut any_attr = false;
+            if let Meta::List(list) = meta {
+                for nested in list.nested.iter() {
+                    if let NestedMeta::Meta(meta) = nested {
+                        if &meta.name().to_string() == "ignore" {
```
Comment (oli-obk) [inline src/librustc_macros/src/hash_stable.rs]: I think instead of the `any_attr` variable you should match on `meta.name().to_string()` and panic for unknown attributes and also just unwrap all the wrong data that can be in one of the two possible attributes. So no `if let` without an `else` in case there are no convenience functions that give you something unwrappable.

File: src/librustc_metadata/creader.rs [hunk 4]
```diff
@@ -281,53 +347,39 @@ impl<'a> CrateLoader<'a> {
         mut dep_kind: DepKind,
     ) -> Result<(CrateNum, Lrc<cstore::CrateMetadata>), LoadError<'b>> {
         info!("resolving crate `extern crate {} as {}`", name, ident);
+        let mut locate_ctxt = locator::Context {
```
Comment (oli-obk) [inline src/librustc_metadata/creader.rs]: why did you move this out? It doesn't seem to be used outside the scope where it used to be in

File: src/librustc_metadata/encoder.rs [hunk 5]
```diff
@@ -490,7 +490,11 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
             } else {
                 None
             },
-
+            proc_macro_stability: if is_proc_macro {
+                tcx.lookup_stability(DefId::local(CRATE_DEF_INDEX)).map(|stab| stab.clone())
```
Comment (oli-obk) [inline src/librustc_metadata/encoder.rs]: please add a test that shows that `HashStable` is unstable

**Review by oli-obk — COMMENTED** [review]
just some nits, then the proc macro part lgtm

File: src/librustc_macros/src/hash_stable.rs (hunk 3 above)
Comment (Zoxc) [reply src/librustc_macros/src/hash_stable.rs]: There are a ton of `if`s here. It's much simpler to just have a panic at the end if nothing matched.

File: src/librustc_metadata/creader.rs (hunk 4 above)
Comment (Zoxc) [reply src/librustc_metadata/creader.rs]: I had to move it out while working on it. I've moved it back now.

File: src/librustc_metadata/encoder.rs (hunk 5 above)
Comment (Zoxc) [reply src/librustc_metadata/encoder.rs]: Added a test.

**Comment by michaelwoerister** [comment]
Looks like this might be good to go now?

File: src/librustc/lib.rs [hunk 6]
```diff
@@ -141,14 +141,8 @@ pub mod util {
...
-// A private module so that macro-expanded idents like
-// `::rustc::lint::Lint` will also work in `rustc` itself.
-//
-// `libstd` uses the same trick.
-#[doc(hidden)]
-mod rustc {
-    pub use crate::lint;
-}
+// Allows macros to refer to this crate as `::rustc`
+extern crate self as rustc;
```
Comment (bjorn3) [inline src/librustc/lib.rs]: Didn't know that was possible :)

**Comment by michaelwoerister** [comment]
:tada:

File: src/librustc_metadata/locator.rs [hunk 7]
```diff
@@ -258,7 +260,7 @@ pub struct Context<'a> {
     pub extra_filename: Option<&'a str>,
     // points to either self.sess.target.target or self.sess.host, must match triple
     pub target: &'a Target,
-    pub triple: &'a TargetTriple,
+    pub triple: TargetTriple,
```
Comment (eddyb) [inline src/librustc_metadata/locator.rs]: cc @nnethercote

File: src/librustc/session/config.rs [hunk 8]
```diff
@@ -1232,6 +1232,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
          Use with RUST_REGION_GRAPH=help for more info"),
     parse_only: bool = (false, parse_bool, [UNTRACKED],
         "parse only; do not compile, assemble, or link"),
+    dual_proc_macros: bool = (false, parse_bool, [TRACKED],
+        "load proc macros for both target and host, but only link to the target"),
```
Comment (eddyb) [inline src/librustc/session/config.rs]: Sorry for being so late to see this, but can an issue be opened to track this "feature"?

Comment (ehuss) [reply src/librustc/session/config.rs]: Opened https://github.com/rust-lang/rust/issues/64671

**Comment by jyn514** [comment]
@Zoxc what's the reason for the last commit ("make cargo a rustc tool again")? It prevents building cargo without building the whole compiler first.

**Comment by eddyb** [comment]
@jyn514 looks like it resolves this failure https://github.com/rust-lang/rust/pull/58013#issuecomment-470092483, and it's just a revert of https://github.com/rust-lang/rust/pull/58013/commits/e501a87e895221dfe7475c6739feffe4bbca04eb in this PR - if you look at [the full diff](https://github.com/rust-lang/rust/pull/58013/files), this PR doesn't change that part of `src/bootstrap/tool.rs` - sounds like you took a wrong turn during a `git blame`?

**Comment by jyn514** [comment]
Hmm ok - I was hoping to remove that dependency in https://github.com/rust-lang/rust/pull/84780#issuecomment-830551841, I guess I should test freebsd locally.

Yes, probably wrong git blame, thanks for the pointer :)

**Comment by eddyb** [comment]
> I guess I should test freebsd locally.

I suspect this will show up on linux too, it looks like some sort of Cargo feature conflict between tools.
